File: P700_Eulercoin.py
'''
TODO:
Bruteforce takes too long. Need to find the inverse of MOD maybe.

Answer is 1517926220024813 considering the first 1e8 multiples of 'x'.

See:
https://urmaul.com/blog/solution-for-project-euler-problem-700/
'''

x = 1504170715041707
MOD = 4503599627370517

min = x
sum = x
print(x, sum, 1)

n = int(1e5)

for i in range(2, n+1):
    next = ((x % MOD) * (i % MOD)) % MOD
    if next < min:
        min = next
        sum += next
        print(next, sum, i)

# 1517926220024813 till 1e8

File: solutions/700-799/P700_Eulercoin.cpp
/*
Question: 

*/

/*
Logic:
There are mistakes in C++ code due to integer overflow. 
Bruteforce takes too long. Need to find inverse of the MOD. 

See:
https://urmaul.com/blog/solution-for-project-euler-problem-700/
*/

// TODO:

#include <iostream>
#include <set>
#include <vector>

typedef unsigned long long ll;

using namespace std;

void solve();

int main() {
    solve();
}

void solve() {
    ll x = 1504170715041707LL;
    constexpr ll MOD =  4503599627370517LL;

    ll min = x;
    ll sum = x;
    ll next;
    cout << x << ' ' << sum << " 1" << '\n';
    ll n = 1e5;
    for (ll i = 2; i <= n; ++i) {
        next = ((x % MOD) * (i % MOD)) % MOD;
        if (next < min) {
            min = next;
            sum += next;
            cout << next << ' ' << sum << ' ' << i << '\n';
        } else {
            // cout << "No\n";
        }
    }
}

// 422691927098 1517440153755132 11117
// 267350072099 1517707503827231 17686
